From 9aa539e209281201256a71e2ebc434f7fd97262b Mon Sep 17 00:00:00 2001
From: Juan Pablo Ugarte <juanpablougarte@gmail.com>
Date: Thu, 5 Sep 2013 15:14:03 -0300
Subject: [PATCH] Fixed bug 702886 "SEGV when creating a new GtkDialog (GLib
 2.37.x)" glib changed the order construct properties are set which exposed
 this bug, now they are set in the order they are installed in the class. So
 we fixed the bug by installing adaptor property first and object last.

---
 gladeui/glade-widget.c | 41 ++++++++++++++++++++++-------------------
 1 file changed, 22 insertions(+), 19 deletions(-)

diff --git a/gladeui/glade-widget.c b/gladeui/glade-widget.c
index 4a974032..d573b5c2 100644
--- a/gladeui/glade-widget.c
+++ b/gladeui/glade-widget.c
@@ -84,11 +84,10 @@ enum
 enum
 {
 	PROP_0,
+	PROP_ADAPTOR,
 	PROP_NAME,
 	PROP_INTERNAL,
 	PROP_ANARCHIST,
-	PROP_OBJECT,
-	PROP_ADAPTOR,
 	PROP_PROJECT,
 	PROP_PROPERTIES,
 	PROP_PARENT,
@@ -98,7 +97,8 @@ enum
 	PROP_REASON,
 	PROP_TOPLEVEL_WIDTH,
 	PROP_TOPLEVEL_HEIGHT,
-	PROP_SUPPORT_WARNING
+	PROP_SUPPORT_WARNING,
+	PROP_OBJECT
 };
 
 static guint         glade_widget_signals[LAST_SIGNAL] = {0};
@@ -1083,6 +1083,14 @@ glade_widget_class_init (GladeWidgetClass *klass)
 	klass->button_release_event   = NULL;
 	klass->motion_notify_event    = NULL;
 
+	g_object_class_install_property
+		(object_class, PROP_ADAPTOR,
+		   g_param_spec_object ("adaptor", _("Adaptor"),
+					_("The class adaptor for the associated widget"),
+					GLADE_TYPE_WIDGET_ADAPTOR,
+					G_PARAM_READWRITE |
+					G_PARAM_CONSTRUCT_ONLY));
+	
 	g_object_class_install_property
 		(object_class, PROP_NAME,
 		 g_param_spec_string ("name", _("Name"),
@@ -1106,22 +1114,6 @@ glade_widget_class_init (GladeWidgetClass *klass)
 				       FALSE, G_PARAM_READWRITE |
 				       G_PARAM_CONSTRUCT_ONLY));
 
-	g_object_class_install_property
-		(object_class, PROP_OBJECT,
-		 g_param_spec_object ("object", _("Object"),
-				      _("The object associated"),
-				      G_TYPE_OBJECT,
-				      G_PARAM_READWRITE |
-				      G_PARAM_CONSTRUCT));
-
-	g_object_class_install_property
-		(object_class, PROP_ADAPTOR,
-		   g_param_spec_object ("adaptor", _("Adaptor"),
-					_("The class adaptor for the associated widget"),
-					GLADE_TYPE_WIDGET_ADAPTOR,
-					G_PARAM_READWRITE |
-					G_PARAM_CONSTRUCT_ONLY));
-
 	g_object_class_install_property
 		(object_class, PROP_PROJECT,
 		 g_param_spec_object ("project", _("Project"),
@@ -1200,6 +1192,17 @@ glade_widget_class_init (GladeWidgetClass *klass)
 				      _("A warning string about version mismatches"),
 				      NULL, G_PARAM_READABLE));
 
+	/* this property setter depends on adaptor and internal properties to be set
+	 * so we install it at the end since in new glib construct properties are
+	 * set in instalation order!
+	 */
+	g_object_class_install_property
+		(object_class, PROP_OBJECT,
+		 g_param_spec_object ("object", _("Object"),
+				      _("The object associated"),
+				      G_TYPE_OBJECT,
+				      G_PARAM_READWRITE));
+
 	/**
 	 * GladeWidget::add-signal-handler:
 	 * @gladewidget: the #GladeWidget which received the signal.
-- 
2.26.2

